// @0no-co/graphqlsp/dist/graphqlsp.mjs
import { C as e, o as t, t as n, g as r, f as i, b as a, a as o, i as s, c as u, d as c, p as l, e as d, r as f, h as v, j as g, k as p, A as m, l as E, m as T } from "./chunks/api-chunk.mjs";

import h from "node:fs/promises";

import y from "path";

import { resolveTypeScriptRootDir as S, parseConfig as I, loadRef as b, getURLConfig as N, minifyIntrospection as D, outputIntrospectionFile as A, extractIntrospectionHeader as F } from "@gql.tada/internal";

import { SchemaMetaFieldDef as k, TypeMetaFieldDef as L, TypeNameMetaFieldDef as _, isCompositeType as O, Kind as x, isScalarType as M, isObjectType as C, isInterfaceType as P, isUnionType as w, isEnumType as R, isInputObjectType as U, getNamedType as j, isOutputType as V, isInputType as $, GraphQLEnumType as Y, GraphQLBoolean as G, isAbstractType as B, assertAbstractType as Q, doTypesOverlap as X, DirectiveLocation as K, visit as J, parse as z, BREAK as H, isListType as W, isNonNullType as q, GraphQLInterfaceType as Z, GraphQLObjectType as ee, GraphQLInputObjectType as te, getNullableType as ne, GraphQLList as re, GraphQLNonNull as ie, TypeInfo as ae, visitWithTypeInfo as oe } from "graphql";

var statFile = (e, t) => h.stat(e).then(t).catch(() => !1);

var swapWrite = async (e, t) => {
  if (!await statFile(e, e => e.isFile())) {
    await h.writeFile(e, t);
  } else {
    var n = e + ".tmp";
    await h.writeFile(n, t);
    try {
      await h.rename(n, e);
    } catch (e) {
      await h.unlink(n);
      throw e;
    } finally {
      await (async e => {
        try {
          var t = new Date;
          await h.utimes(e, t, t);
        } catch (e) {}
      })(e);
    }
  }
};

var toErrorMessage = e => e instanceof Error ? e.message : `${e}`;

var getSchemaForName = (e, t) => t && e.multi[t] ? e.multi[t]?.schema : e.current?.schema;

var se = {};

var loadSchema = (e, t) => {
  var n = {
    config: null,
    load: new Map,
    write: new Map
  };
  var r = new Map;
  var i = new Map;
  var a = new Map;
  var o = null;
  var s = null;
  var u = 0;
  var c = {
    errors: n,
    outputLocations: r,
    sourceLocations: i,
    turboLocations: a,
    get current() {
      return o && o.current;
    },
    get multi() {
      return o ? o.multi : se;
    },
    get version() {
      return o ? o.version : 0;
    },
    checkStale() {
      var e = Date.now();
      if (!s || e - u < 5e3) {
        return;
      }
      u = e;
      s();
    }
  };
  (async () => {
    var u = e.project.getProjectName();
    var c = await S(u) || y.dirname(u);
    t("Got root-directory to resolve schema from: " + c);
    var l;
    try {
      l = I(e.config, c);
    } catch (e) {
      n.config = `${toErrorMessage(e)}. Update the GraphQLSP plugin's entry in "compilerOptions.plugins" in your tsconfig.json.`;
      t(`Found invalid configuration: ${e}`);
      return;
    }
    var d = e.config.tadaDisablePreprocessing ?? !1;
    t('Resolving schema from "schema" config: ' + JSON.stringify(l));
    i.clear();
    a.clear();
    var rememberLocations = (e, t) => {
      if ("string" == typeof t.schema && !N(t.schema)) {
        i.set(e, y.resolve(c, t.schema));
      }
      if (t.tadaTurboLocation) {
        a.set(e, y.resolve(c, t.tadaTurboLocation));
      }
    };
    if ("schemas" in l) {
      for (var f of l.schemas) {
        rememberLocations(f.name, f);
      }
    } else {
      rememberLocations(null, l);
    }
    o = b(l);
    var setLoadError = (e, t) => {
      n.load.set(e || null, `Failed to load the GraphQL schema${e ? ` "${e}"` : ""}: ${toErrorMessage(t)}`);
    };
    var persistSchema = e => {
      if (e && e.tadaOutputLocation) {
        !async function saveTadaIntrospection(e, t, n, r, i, a) {
          try {
            var o = t;
            if (await statFile(o, e => e.isDirectory())) {
              o = y.join(o, "introspection.d.ts");
            } else if (!await statFile(o, e => !!e)) {
              await h.mkdir(y.dirname(o), {
                recursive: !0
              });
              if (await statFile(o, e => e.isDirectory())) {
                o = y.join(o, "introspection.d.ts");
              }
            }
            var s = await h.readFile(o, "utf8").catch(() => {});
            var u = D(e);
            var c = A(u, {
              fileType: t,
              shouldPreprocess: !n,
              preamble: s ? F(s) || void 0 : void 0
            });
            await swapWrite(o, c);
            r.write.delete(t);
            i.set(o, Date.now());
            a(`Introspection saved to path @ ${o}`);
          } catch (e) {
            r.write.set(t, `Failed to write the typings file to "${t}": ${toErrorMessage(e)}`);
            a(`Failed to write introspection @ ${e}`);
          }
        }(e.introspection, y.resolve(c, e.tadaOutputLocation), d, n, r, t);
      }
    };
    var persistAll = () => {
      if (o.current) {
        persistSchema(o.current);
      }
      for (var e in o.multi) {
        persistSchema(o.multi[e]);
      }
    };
    var v = ("schemas" in l ? l.schemas : [ l ]).map(e => e.schema).filter(e => "string" == typeof e && !N(e)).map(e => y.resolve(c, e));
    var g = new Map;
    var recordMtimes = async () => {
      await Promise.all(v.map(async e => {
        try {
          g.set(e, (await h.stat(e)).mtimeMs);
        } catch (e) {}
      }));
    };
    var revalidate = e => o.load({
      rootPath: c,
      reload: e
    }).then(() => {
      n.load.clear();
    }).catch(e => {
      n.load.clear();
      setLoadError(null, e);
    }).then(recordMtimes);
    s = () => {
      (async () => {
        var e = !1;
        for (var n of v) {
          try {
            var r = g.get(n);
            var i = (await h.stat(n)).mtimeMs;
            if (void 0 !== r && i > r) {
              e = !0;
            }
          } catch (e) {}
        }
        if (!e) {
          return;
        }
        t("Schema files changed without watcher events; reloading...");
        await revalidate(!0);
        persistAll();
      })().catch(e => {
        t(`Failed to check schemas for staleness: ${e}`);
      });
    };
    try {
      t("Loading schema...");
      await o.load({
        rootPath: c
      });
      n.load.clear();
    } catch (e) {
      setLoadError(null, e);
      t(`Failed to load schema: ${e}`);
    }
    await recordMtimes();
    persistAll();
    o.autoupdate({
      rootPath: c
    }, (e, t) => {
      revalidate();
      if (!t) {
        return;
      }
      var n = t.name ? e.multi[t.name] : e.current;
      if (n) {
        persistSchema(n);
      }
    }, (e, r) => {
      n.load.delete(null);
      setLoadError(r && r.name, e);
      recordMtimes();
      t(`Failed to load schema${r && r.name ? ` "${r.name}"` : ""} while watching: ${e}`);
    });
  })().catch(e => {
    !function setUnattributedError(e, t) {
      e.load.set(null, `Failed to load the GraphQL schema: ${toErrorMessage(t)}`);
    }(n, e);
    t(`Unexpected error while loading schema: ${e}`);
  });
  return c;
};

function getDefinitionState(e) {
  var t;
  forEachState(e, e => {
    switch (e.kind) {
     case "Query":
     case "ShortQuery":
     case "Mutation":
     case "Subscription":
     case "FragmentDefinition":
      t = e;
    }
  });
  return t;
}

function getFieldDef(e, t, n) {
  if (n === k.name && e.getQueryType() === t) {
    return k;
  }
  if (n === L.name && e.getQueryType() === t) {
    return L;
  }
  if (n === _.name && O(t)) {
    return _;
  }
  if ("getFields" in t) {
    return t.getFields()[n];
  }
  return null;
}

function forEachState(e, t) {
  var n = [];
  var r = e;
  while (null == r ? void 0 : r.kind) {
    n.push(r);
    r = r.prevState;
  }
  for (var i = n.length - 1; i >= 0; i--) {
    t(n[i]);
  }
}

function objectValues(e) {
  var t = Object.keys(e);
  var n = t.length;
  var r = new Array(n);
  for (var i = 0; i < n; ++i) {
    r[i] = e[t[i]];
  }
  return r;
}

function hintList$1(e, t) {
  return function filterAndSortList$1(e, t) {
    if (!t) {
      return filterNonEmpty$1(e, e => !e.isDeprecated);
    }
    var n = e.map(e => ({
      proximity: getProximity$1(normalizeText$1(e.label), t),
      entry: e
    }));
    return filterNonEmpty$1(filterNonEmpty$1(n, e => e.proximity <= 2), e => !e.entry.isDeprecated).sort((e, t) => (e.entry.isDeprecated ? 1 : 0) - (t.entry.isDeprecated ? 1 : 0) || e.proximity - t.proximity || e.entry.label.length - t.entry.label.length).map(e => e.entry);
  }(t, normalizeText$1(e.string));
}

function filterNonEmpty$1(e, t) {
  var n = e.filter(t);
  return 0 === n.length ? e : n;
}

function normalizeText$1(e) {
  return e.toLowerCase().replaceAll(/\W/g, "");
}

function getProximity$1(e, t) {
  var n = function lexicalDistance$1(e, t) {
    var n;
    var r;
    var i = [];
    var a = e.length;
    var o = t.length;
    for (n = 0; n <= a; n++) {
      i[n] = [ n ];
    }
    for (r = 1; r <= o; r++) {
      i[0][r] = r;
    }
    for (n = 1; n <= a; n++) {
      for (r = 1; r <= o; r++) {
        var s = e[n - 1] === t[r - 1] ? 0 : 1;
        i[n][r] = Math.min(i[n - 1][r] + 1, i[n][r - 1] + 1, i[n - 1][r - 1] + s);
        if (n > 1 && r > 1 && e[n - 1] === t[r - 2] && e[n - 2] === t[r - 1]) {
          i[n][r] = Math.min(i[n][r], i[n - 2][r - 2] + s);
        }
      }
    }
    return i[a][o];
  }(t, e);
  if (e.length > t.length) {
    n -= e.length - t.length - 1;
    n += 0 === e.indexOf(t) ? 0 : .5;
  }
  return n;
}

var ue;

!function(e) {
  e.is = function is(e) {
    return "string" == typeof e;
  };
}(ue || (ue = {}));

var ce;

!function(e) {
  e.is = function is(e) {
    return "string" == typeof e;
  };
}(ce || (ce = {}));

var le;

!function(e) {
  e.MIN_VALUE = -2147483648;
  e.MAX_VALUE = 2147483647;
  e.is = function is(t) {
    return "number" == typeof t && e.MIN_VALUE <= t && t <= e.MAX_VALUE;
  };
}(le || (le = {}));

var de;

!function(e) {
  e.MIN_VALUE = 0;
  e.MAX_VALUE = 2147483647;
  e.is = function is(t) {
    return "number" == typeof t && e.MIN_VALUE <= t && t <= e.MAX_VALUE;
  };
}(de || (de = {}));

var fe;

!function(e) {
  e.create = function create(e, t) {
    if (e === Number.MAX_VALUE) {
      e = de.MAX_VALUE;
    }
    if (t === Number.MAX_VALUE) {
      t = de.MAX_VALUE;
    }
    return {
      line: e,
      character: t
    };
  };
  e.is = function is(e) {
    var t = e;
    return Mt.objectLiteral(t) && Mt.uinteger(t.line) && Mt.uinteger(t.character);
  };
}(fe || (fe = {}));

var ve;

!function(e) {
  e.create = function create(e, t, n, r) {
    if (Mt.uinteger(e) && Mt.uinteger(t) && Mt.uinteger(n) && Mt.uinteger(r)) {
      return {
        start: fe.create(e, t),
        end: fe.create(n, r)
      };
    } else if (fe.is(e) && fe.is(t)) {
      return {
        start: e,
        end: t
      };
    } else {
      throw new Error(`Range#create called with invalid arguments[${e}, ${t}, ${n}, ${r}]`);
    }
  };
  e.is = function is(e) {
    var t = e;
    return Mt.objectLiteral(t) && fe.is(t.start) && fe.is(t.end);
  };
}(ve || (ve = {}));

var ge;

!function(e) {
  e.create = function create(e, t) {
    return {
      uri: e,
      range: t
    };
  };
  e.is = function is(e) {
    var t = e;
    return Mt.objectLiteral(t) && ve.is(t.range) && (Mt.string(t.uri) || Mt.undefined(t.uri));
  };
}(ge || (ge = {}));

var pe;

!function(e) {
  e.create = function create(e, t, n, r) {
    return {
      targetUri: e,
      targetRange: t,
      targetSelectionRange: n,
      originSelectionRange: r
    };
  };
  e.is = function is(e) {
    var t = e;
    return Mt.objectLiteral(t) && ve.is(t.targetRange) && Mt.string(t.targetUri) && ve.is(t.targetSelectionRange) && (ve.is(t.originSelectionRange) || Mt.undefined(t.originSelectionRange));
  };
}(pe || (pe = {}));

var me;

!function(e) {
  e.create = function create(e, t, n, r) {
    return {
      red: e,
      green: t,
      blue: n,
      alpha: r
    };
  };
  e.is = function is(e) {
    var t = e;
    return Mt.objectLiteral(t) && Mt.numberRange(t.red, 0, 1) && Mt.numberRange(t.green, 0, 1) && Mt.numberRange(t.blue, 0, 1) && Mt.numberRange(t.alpha, 0, 1);
  };
}(me || (me = {}));

var Ee;

!function(e) {
  e.create = function create(e, t) {
    return {
      range: e,
      color: t
    };
  };
  e.is = function is(e) {
    var t = e;
    return Mt.objectLiteral(t) && ve.is(t.range) && me.is(t.color);
  };
}(Ee || (Ee = {}));

var Te;

!function(e) {
  e.create = function create(e, t, n) {
    return {
      label: e,
      textEdit: t,
      additionalTextEdits: n
    };
  };
  e.is = function is(e) {
    var t = e;
    return Mt.objectLiteral(t) && Mt.string(t.label) && (Mt.undefined(t.textEdit) || Fe.is(t)) && (Mt.undefined(t.additionalTextEdits) || Mt.typedArray(t.additionalTextEdits, Fe.is));
  };
}(Te || (Te = {}));

var he;

!function(e) {
  e.Comment = "comment";
  e.Imports = "imports";
  e.Region = "region";
}(he || (he = {}));

var ye;

!function(e) {
  e.create = function create(e, t, n, r, i, a) {
    var o = {
      startLine: e,
      endLine: t
    };
    if (Mt.defined(n)) {
      o.startCharacter = n;
    }
    if (Mt.defined(r)) {
      o.endCharacter = r;
    }
    if (Mt.defined(i)) {
      o.kind = i;
    }
    if (Mt.defined(a)) {
      o.collapsedText = a;
    }
    return o;
  };
  e.is = function is(e) {
    var t = e;
    return Mt.objectLiteral(t) && Mt.uinteger(t.startLine) && Mt.uinteger(t.startLine) && (Mt.undefined(t.startCharacter) || Mt.uinteger(t.startCharacter)) && (Mt.undefined(t.endCharacter) || Mt.uinteger(t.endCharacter)) && (Mt.undefined(t.kind) || Mt.string(t.kind));
  };
}(ye || (ye = {}));

var Se;

!function(e) {
  e.create = function create(e, t) {
    return {
      location: e,
      message: t
    };
  };
  e.is = function is(e) {
    var t = e;
    return Mt.defined(t) && ge.is(t.location) && Mt.string(t.message);
  };
}(Se || (Se = {}));

var Ie;

!function(e) {
  e.Error = 1;
  e.Warning = 2;
  e.Information = 3;
  e.Hint = 4;
}(Ie || (Ie = {}));

var be;

!function(e) {
  e.Unnecessary = 1;
  e.Deprecated = 2;
}(be || (be = {}));

var Ne;

!function(e) {
  e.is = function is(e) {
    var t = e;
    return Mt.objectLiteral(t) && Mt.string(t.href);
  };
}(Ne || (Ne = {}));

var De;

!function(e) {
  e.create = function create(e, t, n, r, i, a) {
    var o = {
      range: e,
      message: t
    };
    if (Mt.defined(n)) {
      o.severity = n;
    }
    if (Mt.defined(r)) {
      o.code = r;
    }
    if (Mt.defined(i)) {
      o.source = i;
    }
    if (Mt.defined(a)) {
      o.relatedInformation = a;
    }
    return o;
  };
  e.is = function is(e) {
    var t;
    var n = e;
    return Mt.defined(n) && ve.is(n.range) && Mt.string(n.message) && (Mt.number(n.severity) || Mt.undefined(n.severity)) && (Mt.integer(n.code) || Mt.string(n.code) || Mt.undefined(n.code)) && (Mt.undefined(n.codeDescription) || Mt.string(null === (t = n.codeDescription) || void 0 === t ? void 0 : t.href)) && (Mt.string(n.source) || Mt.undefined(n.source)) && (Mt.undefined(n.relatedInformation) || Mt.typedArray(n.relatedInformation, Se.is));
  };
}(De || (De = {}));

var Ae;

!function(e) {
  e.create = function create(e, t, ...n) {
    var r = {
      title: e,
      command: t
    };
    if (Mt.defined(n) && n.length > 0) {
      r.arguments = n;
    }
    return r;
  };
  e.is = function is(e) {
    var t = e;
    return Mt.defined(t) && Mt.string(t.title) && Mt.string(t.command);
  };
}(Ae || (Ae = {}));

var Fe;

!function(e) {
  e.replace = function replace(e, t) {
    return {
      range: e,
      newText: t
    };
  };
  e.insert = function insert(e, t) {
    return {
      range: {
        start: e,
        end: e
      },
      newText: t
    };
  };
  e.del = function del(e) {
    return {
      range: e,
      newText: ""
    };
  };
  e.is = function is(e) {
    var t = e;
    return Mt.objectLiteral(t) && Mt.string(t.newText) && ve.is(t.range);
  };
}(Fe || (Fe = {}));

var ke;

!function(e) {
  e.create = function create(e, t, n) {
    var r = {
      label: e
    };
    if (void 0 !== t) {
      r.needsConfirmation = t;
    }
    if (void 0 !== n) {
      r.description = n;
    }
    return r;
  };
  e.is = function is(e) {
    var t = e;
    return Mt.objectLiteral(t) && Mt.string(t.label) && (Mt.boolean(t.needsConfirmation) || void 0 === t.needsConfirmation) && (Mt.string(t.description) || void 0 === t.description);
  };
}(ke || (ke = {}));

var Le;

!function(e) {
  e.is = function is(e) {
    return Mt.string(e);
  };
}(Le || (Le = {}));

var _e;

!function(e) {
  e.replace = function replace(e, t, n) {
    return {
      range: e,
      newText: t,
      annotationId: n
    };
  };
  e.insert = function insert(e, t, n) {
    return {
      range: {
        start: e,
        end: e
      },
      newText: t,
      annotationId: n
    };
  };
  e.del = function del(e, t) {
    return {
      range: e,
      newText: "",
      annotationId: t
    };
  };
  e.is = function is(e) {
    var t = e;
    return Fe.is(t) && (ke.is(t.annotationId) || Le.is(t.annotationId));
  };
}(_e || (_e = {}));

var Oe;

!function(e) {
  e.create = function create(e, t) {
    return {
      textDocument: e,
      edits: t
    };
  };
  e.is = function is(e) {
    var t = e;
    return Mt.defined(t) && Ue.is(t.textDocument) && Array.isArray(t.edits);
  };
}(Oe || (Oe = {}));

var xe;

!function(e) {
  e.create = function create(e, t, n) {
    var r = {
      kind: "create",
      uri: e
    };
    if (void 0 !== t && (void 0 !== t.overwrite || void 0 !== t.ignoreIfExists)) {
      r.options = t;
    }
    if (void 0 !== n) {
      r.annotationId = n;
    }
    return r;
  };
  e.is = function is(e) {
    var t = e;
    return t && "create" === t.kind && Mt.string(t.uri) && (void 0 === t.options || (void 0 === t.options.overwrite || Mt.boolean(t.options.overwrite)) && (void 0 === t.options.ignoreIfExists || Mt.boolean(t.options.ignoreIfExists))) && (void 0 === t.annotationId || Le.is(t.annotationId));
  };
}(xe || (xe = {}));

var Me;

!function(e) {
  e.create = function create(e, t, n, r) {
    var i = {
      kind: "rename",
      oldUri: e,
      newUri: t
    };
    if (void 0 !== n && (void 0 !== n.overwrite || void 0 !== n.ignoreIfExists)) {
      i.options = n;
    }
    if (void 0 !== r) {
      i.annotationId = r;
    }
    return i;
  };
  e.is = function is(e) {
    var t = e;
    return t && "rename" === t.kind && Mt.string(t.oldUri) && Mt.string(t.newUri) && (void 0 === t.options || (void 0 === t.options.overwrite || Mt.boolean(t.options.overwrite)) && (void 0 === t.options.ignoreIfExists || Mt.boolean(t.options.ignoreIfExists))) && (void 0 === t.annotationId || Le.is(t.annotationId));
  };
}(Me || (Me = {}));

var Ce;

!function(e) {
  e.create = function create(e, t, n) {
    var r = {
      kind: "delete",
      uri: e
    };
    if (void 0 !== t && (void 0 !== t.recursive || void 0 !== t.ignoreIfNotExists)) {
      r.options = t;
    }
    if (void 0 !== n) {
      r.annotationId = n;
    }
    return r;
  };
  e.is = function is(e) {
    var t = e;
    return t && "delete" === t.kind && Mt.string(t.uri) && (void 0 === t.options || (void 0 === t.options.recursive || Mt.boolean(t.options.recursive)) && (void 0 === t.options.ignoreIfNotExists || Mt.boolean(t.options.ignoreIfNotExists))) && (void 0 === t.annotationId || Le.is(t.annotationId));
  };
}(Ce || (Ce = {}));

var Pe;

!function(e) {
  e.is = function is(e) {
    return e && (void 0 !== e.changes || void 0 !== e.documentChanges) && (void 0 === e.documentChanges || e.documentChanges.every(e => {
      if (Mt.string(e.kind)) {
        return xe.is(e) || Me.is(e) || Ce.is(e);
      } else {
        return Oe.is(e);
      }
    }));
  };
}(Pe || (Pe = {}));

var we;

!function(e) {
  e.create = function create(e) {
    return {
      uri: e
    };
  };
  e.is = function is(e) {
    var t = e;
    return Mt.defined(t) && Mt.string(t.uri);
  };
}(we || (we = {}));

var Re;

!function(e) {
  e.create = function create(e, t) {
    return {
      uri: e,
      version: t
    };
  };
  e.is = function is(e) {
    var t = e;
    return Mt.defined(t) && Mt.string(t.uri) && Mt.integer(t.version);
  };
}(Re || (Re = {}));

var Ue;

!function(e) {
  e.create = function create(e, t) {
    return {
      uri: e,
      version: t
    };
  };
  e.is = function is(e) {
    var t = e;
    return Mt.defined(t) && Mt.string(t.uri) && (null === t.version || Mt.integer(t.version));
  };
}(Ue || (Ue = {}));

var je;

!function(e) {
  e.create = function create(e, t, n, r) {
    return {
      uri: e,
      languageId: t,
      version: n,
      text: r
    };
  };
  e.is = function is(e) {
    var t = e;
    return Mt.defined(t) && Mt.string(t.uri) && Mt.string(t.languageId) && Mt.integer(t.version) && Mt.string(t.text);
  };
}(je || (je = {}));

var Ve;

!function(e) {
  e.PlainText = "plaintext";
  e.Markdown = "markdown";
  e.is = function is(t) {
    return t === e.PlainText || t === e.Markdown;
  };
}(Ve || (Ve = {}));

var $e;

!function(e) {
  e.is = function is(e) {
    var t = e;
    return Mt.objectLiteral(e) && Ve.is(t.kind) && Mt.string(t.value);
  };
}($e || ($e = {}));

var Ye;

!function(e) {
  e.Text = 1;
  e.Method = 2;
  e.Function = 3;
  e.Constructor = 4;
  e.Field = 5;
  e.Variable = 6;
  e.Class = 7;
  e.Interface = 8;
  e.Module = 9;
  e.Property = 10;
  e.Unit = 11;
  e.Value = 12;
  e.Enum = 13;
  e.Keyword = 14;
  e.Snippet = 15;
  e.Color = 16;
  e.File = 17;
  e.Reference = 18;
  e.Folder = 19;
  e.EnumMember = 20;
  e.Constant = 21;
  e.Struct = 22;
  e.Event = 23;
  e.Operator = 24;
  e.TypeParameter = 25;
}(Ye || (Ye = {}));

var Ge;

!function(e) {
  e.PlainText = 1;
  e.Snippet = 2;
}(Ge || (Ge = {}));

var Be;

!function(e) {
  e.Deprecated = 1;
}(Be || (Be = {}));

var Qe;

!function(e) {
  e.create = function create(e, t, n) {
    return {
      newText: e,
      insert: t,
      replace: n
    };
  };
  e.is = function is(e) {
    var t = e;
    return t && Mt.string(t.newText) && ve.is(t.insert) && ve.is(t.replace);
  };
}(Qe || (Qe = {}));

var Xe;

!function(e) {
  e.asIs = 1;
  e.adjustIndentation = 2;
}(Xe || (Xe = {}));

var Ke;

!function(e) {
  e.is = function is(e) {
    var t = e;
    return t && (Mt.string(t.detail) || void 0 === t.detail) && (Mt.string(t.description) || void 0 === t.description);
  };
}(Ke || (Ke = {}));

var Je;

!function(e) {
  e.create = function create(e) {
    return {
      label: e
    };
  };
}(Je || (Je = {}));

var ze;

!function(e) {
  e.create = function create(e, t) {
    return {
      items: e ? e : [],
      isIncomplete: !!t
    };
  };
}(ze || (ze = {}));

var He;

!function(e) {
  e.fromPlainText = function fromPlainText(e) {
    return e.replace(/[\\`*_{}[\]()#+\-.!]/g, "\\$&");
  };
  e.is = function is(e) {
    var t = e;
    return Mt.string(t) || Mt.objectLiteral(t) && Mt.string(t.language) && Mt.string(t.value);
  };
}(He || (He = {}));

var We;

!function(e) {
  e.is = function is(e) {
    var t = e;
    return !!t && Mt.objectLiteral(t) && ($e.is(t.contents) || He.is(t.contents) || Mt.typedArray(t.contents, He.is)) && (void 0 === e.range || ve.is(e.range));
  };
}(We || (We = {}));

var qe;

!function(e) {
  e.create = function create(e, t) {
    return t ? {
      label: e,
      documentation: t
    } : {
      label: e
    };
  };
}(qe || (qe = {}));

var Ze;

!function(e) {
  e.create = function create(e, t, ...n) {
    var r = {
      label: e
    };
    if (Mt.defined(t)) {
      r.documentation = t;
    }
    if (Mt.defined(n)) {
      r.parameters = n;
    } else {
      r.parameters = [];
    }
    return r;
  };
}(Ze || (Ze = {}));

var et;

!function(e) {
  e.Text = 1;
  e.Read = 2;
  e.Write = 3;
}(et || (et = {}));

var tt;

!function(e) {
  e.create = function create(e, t) {
    var n = {
      range: e
    };
    if (Mt.number(t)) {
      n.kind = t;
    }
    return n;
  };
}(tt || (tt = {}));

var nt;

!function(e) {
  e.File = 1;
  e.Module = 2;
  e.Namespace = 3;
  e.Package = 4;
  e.Class = 5;
  e.Method = 6;
  e.Property = 7;
  e.Field = 8;
  e.Constructor = 9;
  e.Enum = 10;
  e.Interface = 11;
  e.Function = 12;
  e.Variable = 13;
  e.Constant = 14;
  e.String = 15;
  e.Number = 16;
  e.Boolean = 17;
  e.Array = 18;
  e.Object = 19;
  e.Key = 20;
  e.Null = 21;
  e.EnumMember = 22;
  e.Struct = 23;
  e.Event = 24;
  e.Operator = 25;
  e.TypeParameter = 26;
}(nt || (nt = {}));

var rt;

!function(e) {
  e.Deprecated = 1;
}(rt || (rt = {}));

var it;

!function(e) {
  e.create = function create(e, t, n, r, i) {
    var a = {
      name: e,
      kind: t,
      location: {
        uri: r,
        range: n
      }
    };
    if (i) {
      a.containerName = i;
    }
    return a;
  };
}(it || (it = {}));

var at;

!function(e) {
  e.create = function create(e, t, n, r) {
    return void 0 !== r ? {
      name: e,
      kind: t,
      location: {
        uri: n,
        range: r
      }
    } : {
      name: e,
      kind: t,
      location: {
        uri: n
      }
    };
  };
}(at || (at = {}));

var ot;

!function(e) {
  e.create = function create(e, t, n, r, i, a) {
    var o = {
      name: e,
      detail: t,
      kind: n,
      range: r,
      selectionRange: i
    };
    if (void 0 !== a) {
      o.children = a;
    }
    return o;
  };
  e.is = function is(e) {
    var t = e;
    return t && Mt.string(t.name) && Mt.number(t.kind) && ve.is(t.range) && ve.is(t.selectionRange) && (void 0 === t.detail || Mt.string(t.detail)) && (void 0 === t.deprecated || Mt.boolean(t.deprecated)) && (void 0 === t.children || Array.isArray(t.children)) && (void 0 === t.tags || Array.isArray(t.tags));
  };
}(ot || (ot = {}));

var st;

!function(e) {
  e.Empty = "";
  e.QuickFix = "quickfix";
  e.Refactor = "refactor";
  e.RefactorExtract = "refactor.extract";
  e.RefactorInline = "refactor.inline";
  e.RefactorRewrite = "refactor.rewrite";
  e.Source = "source";
  e.SourceOrganizeImports = "source.organizeImports";
  e.SourceFixAll = "source.fixAll";
}(st || (st = {}));

var ut;

!function(e) {
  e.Invoked = 1;
  e.Automatic = 2;
}(ut || (ut = {}));

var ct;

!function(e) {
  e.create = function create(e, t, n) {
    var r = {
      diagnostics: e
    };
    if (null != t) {
      r.only = t;
    }
    if (null != n) {
      r.triggerKind = n;
    }
    return r;
  };
  e.is = function is(e) {
    var t = e;
    return Mt.defined(t) && Mt.typedArray(t.diagnostics, De.is) && (void 0 === t.only || Mt.typedArray(t.only, Mt.string)) && (void 0 === t.triggerKind || t.triggerKind === ut.Invoked || t.triggerKind === ut.Automatic);
  };
}(ct || (ct = {}));

var lt;

!function(e) {
  e.create = function create(e, t, n) {
    var r = {
      title: e
    };
    var i = !0;
    if ("string" == typeof t) {
      i = !1;
      r.kind = t;
    } else if (Ae.is(t)) {
      r.command = t;
    } else {
      r.edit = t;
    }
    if (i && void 0 !== n) {
      r.kind = n;
    }
    return r;
  };
  e.is = function is(e) {
    var t = e;
    return t && Mt.string(t.title) && (void 0 === t.diagnostics || Mt.typedArray(t.diagnostics, De.is)) && (void 0 === t.kind || Mt.string(t.kind)) && (void 0 !== t.edit || void 0 !== t.command) && (void 0 === t.command || Ae.is(t.command)) && (void 0 === t.isPreferred || Mt.boolean(t.isPreferred)) && (void 0 === t.edit || Pe.is(t.edit));
  };
}(lt || (lt = {}));

var dt;

!function(e) {
  e.create = function create(e, t) {
    var n = {
      range: e
    };
    if (Mt.defined(t)) {
      n.data = t;
    }
    return n;
  };
  e.is = function is(e) {
    var t = e;
    return Mt.defined(t) && ve.is(t.range) && (Mt.undefined(t.command) || Ae.is(t.command));
  };
}(dt || (dt = {}));

var ft;

!function(e) {
  e.create = function create(e, t) {
    return {
      tabSize: e,
      insertSpaces: t
    };
  };
  e.is = function is(e) {
    var t = e;
    return Mt.defined(t) && Mt.uinteger(t.tabSize) && Mt.boolean(t.insertSpaces);
  };
}(ft || (ft = {}));

var vt;

!function(e) {
  e.create = function create(e, t, n) {
    return {
      range: e,
      target: t,
      data: n
    };
  };
  e.is = function is(e) {
    var t = e;
    return Mt.defined(t) && ve.is(t.range) && (Mt.undefined(t.target) || Mt.string(t.target));
  };
}(vt || (vt = {}));

var gt;

!function(e) {
  e.create = function create(e, t) {
    return {
      range: e,
      parent: t
    };
  };
  e.is = function is(t) {
    var n = t;
    return Mt.objectLiteral(n) && ve.is(n.range) && (void 0 === n.parent || e.is(n.parent));
  };
}(gt || (gt = {}));

var pt;

!function(e) {
  e.namespace = "namespace";
  e.type = "type";
  e.class = "class";
  e.enum = "enum";
  e.interface = "interface";
  e.struct = "struct";
  e.typeParameter = "typeParameter";
  e.parameter = "parameter";
  e.variable = "variable";
  e.property = "property";
  e.enumMember = "enumMember";
  e.event = "event";
  e.function = "function";
  e.method = "method";
  e.macro = "macro";
  e.keyword = "keyword";
  e.modifier = "modifier";
  e.comment = "comment";
  e.string = "string";
  e.number = "number";
  e.regexp = "regexp";
  e.operator = "operator";
  e.decorator = "decorator";
}(pt || (pt = {}));

var mt;

!function(e) {
  e.declaration = "declaration";
  e.definition = "definition";
  e.readonly = "readonly";
  e.static = "static";
  e.deprecated = "deprecated";
  e.abstract = "abstract";
  e.async = "async";
  e.modification = "modification";
  e.documentation = "documentation";
  e.defaultLibrary = "defaultLibrary";
}(mt || (mt = {}));

var Et;

!function(e) {
  e.is = function is(e) {
    var t = e;
    return Mt.objectLiteral(t) && (void 0 === t.resultId || "string" == typeof t.resultId) && Array.isArray(t.data) && (0 === t.data.length || "number" == typeof t.data[0]);
  };
}(Et || (Et = {}));

var Tt;

!function(e) {
  e.create = function create(e, t) {
    return {
      range: e,
      text: t
    };
  };
  e.is = function is(e) {
    var t = e;
    return null != t && ve.is(t.range) && Mt.string(t.text);
  };
}(Tt || (Tt = {}));

var ht;

!function(e) {
  e.create = function create(e, t, n) {
    return {
      range: e,
      variableName: t,
      caseSensitiveLookup: n
    };
  };
  e.is = function is(e) {
    var t = e;
    return null != t && ve.is(t.range) && Mt.boolean(t.caseSensitiveLookup) && (Mt.string(t.variableName) || void 0 === t.variableName);
  };
}(ht || (ht = {}));

var yt;

!function(e) {
  e.create = function create(e, t) {
    return {
      range: e,
      expression: t
    };
  };
  e.is = function is(e) {
    var t = e;
    return null != t && ve.is(t.range) && (Mt.string(t.expression) || void 0 === t.expression);
  };
}(yt || (yt = {}));

var St;

!function(e) {
  e.create = function create(e, t) {
    return {
      frameId: e,
      stoppedLocation: t
    };
  };
  e.is = function is(e) {
    return Mt.defined(e) && ve.is(e.stoppedLocation);
  };
}(St || (St = {}));

var It;

!function(e) {
  e.Type = 1;
  e.Parameter = 2;
  e.is = function is(e) {
    return 1 === e || 2 === e;
  };
}(It || (It = {}));

var bt;

!function(e) {
  e.create = function create(e) {
    return {
      value: e
    };
  };
  e.is = function is(e) {
    var t = e;
    return Mt.objectLiteral(t) && (void 0 === t.tooltip || Mt.string(t.tooltip) || $e.is(t.tooltip)) && (void 0 === t.location || ge.is(t.location)) && (void 0 === t.command || Ae.is(t.command));
  };
}(bt || (bt = {}));

var Nt;

!function(e) {
  e.create = function create(e, t, n) {
    var r = {
      position: e,
      label: t
    };
    if (void 0 !== n) {
      r.kind = n;
    }
    return r;
  };
  e.is = function is(e) {
    var t = e;
    return Mt.objectLiteral(t) && fe.is(t.position) && (Mt.string(t.label) || Mt.typedArray(t.label, bt.is)) && (void 0 === t.kind || It.is(t.kind)) && void 0 === t.textEdits || Mt.typedArray(t.textEdits, Fe.is) && (void 0 === t.tooltip || Mt.string(t.tooltip) || $e.is(t.tooltip)) && (void 0 === t.paddingLeft || Mt.boolean(t.paddingLeft)) && (void 0 === t.paddingRight || Mt.boolean(t.paddingRight));
  };
}(Nt || (Nt = {}));

var Dt;

!function(e) {
  e.createSnippet = function createSnippet(e) {
    return {
      kind: "snippet",
      value: e
    };
  };
}(Dt || (Dt = {}));

var At;

!function(e) {
  e.create = function create(e, t, n, r) {
    return {
      insertText: e,
      filterText: t,
      range: n,
      command: r
    };
  };
}(At || (At = {}));

var Ft;

!function(e) {
  e.create = function create(e) {
    return {
      items: e
    };
  };
}(Ft || (Ft = {}));

var kt;

!function(e) {
  e.Invoked = 0;
  e.Automatic = 1;
}(kt || (kt = {}));

var Lt;

!function(e) {
  e.create = function create(e, t) {
    return {
      range: e,
      text: t
    };
  };
}(Lt || (Lt = {}));

var _t;

!function(e) {
  e.create = function create(e, t) {
    return {
      triggerKind: e,
      selectedCompletionInfo: t
    };
  };
}(_t || (_t = {}));

var Ot;

!function(e) {
  e.is = function is(e) {
    var t = e;
    return Mt.objectLiteral(t) && ce.is(t.uri) && Mt.string(t.name);
  };
}(Ot || (Ot = {}));

var xt;

!function(e) {
  e.create = function create(e, t, n, r) {
    return new FullTextDocument(e, t, n, r);
  };
  e.is = function is(e) {
    var t = e;
    return Mt.defined(t) && Mt.string(t.uri) && (Mt.undefined(t.languageId) || Mt.string(t.languageId)) && Mt.uinteger(t.lineCount) && Mt.func(t.getText) && Mt.func(t.positionAt) && Mt.func(t.offsetAt) ? !0 : !1;
  };
  e.applyEdits = function applyEdits(e, t) {
    var n = e.getText();
    var r = mergeSort(t, (e, t) => {
      var n = e.range.start.line - t.range.start.line;
      if (0 === n) {
        return e.range.start.character - t.range.start.character;
      }
      return n;
    });
    var i = n.length;
    for (var a = r.length - 1; a >= 0; a--) {
      var o = r[a];
      var s = e.offsetAt(o.range.start);
      var u = e.offsetAt(o.range.end);
      if (u <= i) {
        n = n.substring(0, s) + o.newText + n.substring(u, n.length);
      } else {
        throw new Error("Overlapping edit");
      }
      i = s;
    }
    return n;
  };
  function mergeSort(e, t) {
    if (e.length <= 1) {
      return e;
    }
    var n = e.length / 2 | 0;
    var r = e.slice(0, n);
    var i = e.slice(n);
    mergeSort(r, t);
    mergeSort(i, t);
    var a = 0;
    var o = 0;
    var s = 0;
    while (a < r.length && o < i.length) {
      if (t(r[a], i[o]) <= 0) {
        e[s++] = r[a++];
      } else {
        e[s++] = i[o++];
      }
    }
    while (a < r.length) {
      e[s++] = r[a++];
    }
    while (o < i.length) {
      e[s++] = i[o++];
    }
    return e;
  }
}(xt || (xt = {}));

class FullTextDocument {
  constructor(e, t, n, r) {
    this._uri = e;
    this._languageId = t;
    this._version = n;
    this._content = r;
    this._lineOffsets = void 0;
  }
  get uri() {
    return this._uri;
  }
  get languageId() {
    return this._languageId;
  }
  get version() {
    return this._version;
  }
  getText(e) {
    if (e) {
      var t = this.offsetAt(e.start);
      var n = this.offsetAt(e.end);
      return this._content.substring(t, n);
    }
    return this._content;
  }
  update(e, t) {
    this._content = e.text;
    this._version = t;
    this._lineOffsets = void 0;
  }
  getLineOffsets() {
    if (void 0 === this._lineOffsets) {
      var e = [];
      var t = this._content;
      var n = !0;
      for (var r = 0; r < t.length; r++) {
        if (n) {
          e.push(r);
          n = !1;
        }
        var i = t.charAt(r);
        n = "\r" === i || "\n" === i;
        if ("\r" === i && r + 1 < t.length && "\n" === t.charAt(r + 1)) {
          r++;
        }
      }
      if (n && t.length > 0) {
        e.push(t.length);
      }
      this._lineOffsets = e;
    }
    return this._lineOffsets;
  }
  positionAt(e) {
    e = Math.max(Math.min(e, this._content.length), 0);
    var t = this.getLineOffsets();
    var n = 0, r = t.length;
    if (0 === r) {
      return fe.create(0, e);
    }
    while (n < r) {
      var i = Math.floor((n + r) / 2);
      if (t[i] > e) {
        r = i;
      } else {
        n = i + 1;
      }
    }
    var a = n - 1;
    return fe.create(a, e - t[a]);
  }
  offsetAt(e) {
    var t = this.getLineOffsets();
    if (e.line >= t.length) {
      return this._content.length;
    } else if (e.line < 0) {
      return 0;
    }
    var n = t[e.line];
    return Math.max(Math.min(n + e.character, e.line + 1 < t.length ? t[e.line + 1] : this._content.length), n);
  }
  get lineCount() {
    return this.getLineOffsets().length;
  }
}

var Mt;

!function(e) {
  var t = Object.prototype.toString;
  e.defined = function defined(e) {
    return void 0 !== e;
  };
  e.undefined = function undefined$1(e) {
    return void 0 === e;
  };
  e.boolean = function boolean(e) {
    return !0 === e || !1 === e;
  };
  e.string = function string(e) {
    return "[object String]" === t.call(e);
  };
  e.number = function number(e) {
    return "[object Number]" === t.call(e);
  };
  e.numberRange = function numberRange(e, n, r) {
    return "[object Number]" === t.call(e) && n <= e && e <= r;
  };
  e.integer = function integer(e) {
    return "[object Number]" === t.call(e) && -2147483648 <= e && e <= 2147483647;
  };
  e.uinteger = function uinteger(e) {
    return "[object Number]" === t.call(e) && 0 <= e && e <= 2147483647;
  };
  e.func = function func(e) {
    return "[object Function]" === t.call(e);
  };
  e.objectLiteral = function objectLiteral(e) {
    return null !== e && "object" == typeof e;
  };
  e.typedArray = function typedArray(e, t) {
    return Array.isArray(e) && e.every(t);
  };
}(Mt || (Mt = {}));

var Ct;

!function(e) {
  e.Text = 1;
  e.Method = 2;
  e.Function = 3;
  e.Constructor = 4;
  e.Field = 5;
  e.Variable = 6;
  e.Class = 7;
  e.Interface = 8;
  e.Module = 9;
  e.Property = 10;
  e.Unit = 11;
  e.Value = 12;
  e.Enum = 13;
  e.Keyword = 14;
  e.Snippet = 15;
  e.Color = 16;
  e.File = 17;
  e.Reference = 18;
  e.Folder = 19;
  e.EnumMember = 20;
  e.Constant = 21;
  e.Struct = 22;
  e.Event = 23;
  e.Operator = 24;
  e.TypeParameter = 25;
}(Ct || (Ct = {}));

var Pt = Object.assign(Object.assign({}, x), {
  ALIASED_FIELD: "AliasedField",
  ARGUMENTS: "Arguments",
  SHORT_QUERY: "ShortQuery",
  QUERY: "Query",
  MUTATION: "Mutation",
  SUBSCRIPTION: "Subscription",
  TYPE_CONDITION: "TypeCondition",
  INVALID: "Invalid",
  COMMENT: "Comment",
  SCHEMA_DEF: "SchemaDef",
  SCALAR_DEF: "ScalarDef",
  OBJECT_TYPE_DEF: "ObjectTypeDef",
  OBJECT_VALUE: "ObjectValue",
  LIST_VALUE: "ListValue",
  INTERFACE_DEF: "InterfaceDef",
  UNION_DEF: "UnionDef",
  ENUM_DEF: "EnumDef",
  ENUM_VALUE: "EnumValue",
  FIELD_DEF: "FieldDef",
  INPUT_DEF: "InputDef",
  INPUT_VALUE_DEF: "InputValueDef",
  ARGUMENTS_DEF: "ArgumentsDef",
  EXTEND_DEF: "ExtendDef",
  EXTENSION_DEFINITION: "ExtensionDefinition",
  DIRECTIVE_DEF: "DirectiveDef",
  IMPLEMENTS: "Implements",
  VARIABLE_DEFINITIONS: "VariableDefinitions",
  TYPE: "Type"
});

var wt = {
  command: "editor.action.triggerSuggest",
  title: "Suggestions"
};

var Rt = [ x.SCHEMA_DEFINITION, x.OPERATION_TYPE_DEFINITION, x.SCALAR_TYPE_DEFINITION, x.OBJECT_TYPE_DEFINITION, x.INTERFACE_TYPE_DEFINITION, x.UNION_TYPE_DEFINITION, x.ENUM_TYPE_DEFINITION, x.INPUT_OBJECT_TYPE_DEFINITION, x.DIRECTIVE_DEFINITION, x.SCHEMA_EXTENSION, x.SCALAR_TYPE_EXTENSION, x.OBJECT_TYPE_EXTENSION, x.INTERFACE_TYPE_EXTENSION, x.UNION_TYPE_EXTENSION, x.ENUM_TYPE_EXTENSION, x.INPUT_OBJECT_TYPE_EXTENSION ];

function getAutocompleteSuggestions(e, t, n, r, i, a) {
  var o;
  var s = Object.assign(Object.assign({}, a), {
    schema: e
  });
  var u = r || getTokenAtPosition(t, n, 1);
  var c = "Invalid" === u.state.kind ? u.state.prevState : u.state;
  var l = (null == a ? void 0 : a.mode) || function getDocumentMode(e, t) {
    if (null == t ? void 0 : t.endsWith(".graphqls")) {
      return jt.TYPE_SYSTEM;
    }
    return (e => {
      var t = !1;
      if (e) {
        try {
          J(z(e), {
            enter(e) {
              if ("Document" === e.kind) {
                return;
              }
              if (Rt.includes(e.kind)) {
                t = !0;
                return H;
              }
              return !1;
            }
          });
        } catch (e) {
          return t;
        }
      }
      return t;
    })(e) ? jt.TYPE_SYSTEM : jt.EXECUTABLE;
  }(t, null == a ? void 0 : a.uri);
  if (!c) {
    return [];
  }
  var {kind: d, step: f, prevState: v} = c;
  var g = getTypeInfo(e, u.state);
  if (d === Pt.DOCUMENT) {
    if (l === jt.TYPE_SYSTEM) {
      return function getSuggestionsForTypeSystemDefinitions(e) {
        return hintList$1(e, [ {
          label: "extend",
          kind: Ct.Function
        }, {
          label: "type",
          kind: Ct.Function
        }, {
          label: "interface",
          kind: Ct.Function
        }, {
          label: "union",
          kind: Ct.Function
        }, {
          label: "input",
          kind: Ct.Function
        }, {
          label: "scalar",
          kind: Ct.Function
        }, {
          label: "schema",
          kind: Ct.Function
        } ]);
      }(u);
    }
    return function getSuggestionsForExecutableDefinitions(e) {
      return hintList$1(e, [ {
        label: "query",
        kind: Ct.Function
      }, {
        label: "mutation",
        kind: Ct.Function
      }, {
        label: "subscription",
        kind: Ct.Function
      }, {
        label: "fragment",
        kind: Ct.Function
      }, {
        label: "{",
        kind: Ct.Constructor
      } ]);
    }(u);
  }
  if (d === Pt.EXTEND_DEF) {
    return function getSuggestionsForExtensionDefinitions(e) {
      return hintList$1(e, [ {
        label: "type",
        kind: Ct.Function
      }, {
        label: "interface",
        kind: Ct.Function
      }, {
        label: "union",
        kind: Ct.Function
      }, {
        label: "input",
        kind: Ct.Function
      }, {
        label: "scalar",
        kind: Ct.Function
      }, {
        label: "schema",
        kind: Ct.Function
      } ]);
    }(u);
  }
  if ((null === (o = null == v ? void 0 : v.prevState) || void 0 === o ? void 0 : o.kind) === Pt.EXTENSION_DEFINITION && c.name) {
    return hintList$1(u, []);
  }
  if ((null == v ? void 0 : v.kind) === x.SCALAR_TYPE_EXTENSION) {
    return hintList$1(u, Object.values(e.getTypeMap()).filter(M).map(e => ({
      label: e.name,
      kind: Ct.Function
    })));
  }
  if ((null == v ? void 0 : v.kind) === x.OBJECT_TYPE_EXTENSION) {
    return hintList$1(u, Object.values(e.getTypeMap()).filter(e => C(e) && !e.name.startsWith("__")).map(e => ({
      label: e.name,
      kind: Ct.Function
    })));
  }
  if ((null == v ? void 0 : v.kind) === x.INTERFACE_TYPE_EXTENSION) {
    return hintList$1(u, Object.values(e.getTypeMap()).filter(P).map(e => ({
      label: e.name,
      kind: Ct.Function
    })));
  }
  if ((null == v ? void 0 : v.kind) === x.UNION_TYPE_EXTENSION) {
    return hintList$1(u, Object.values(e.getTypeMap()).filter(w).map(e => ({
      label: e.name,
      kind: Ct.Function
    })));
  }
  if ((null == v ? void 0 : v.kind) === x.ENUM_TYPE_EXTENSION) {
    return hintList$1(u, Object.values(e.getTypeMap()).filter(e => R(e) && !e.name.startsWith("__")).map(e => ({
      label: e.name,
      kind: Ct.Function
    })));
  }
  if ((null == v ? void 0 : v.kind) === x.INPUT_OBJECT_TYPE_EXTENSION) {
    return hintList$1(u, Object.values(e.getTypeMap()).filter(U).map(e => ({
      label: e.name,
      kind: Ct.Function
    })));
  }
  if (d === Pt.IMPLEMENTS || d === Pt.NAMED_TYPE && (null == v ? void 0 : v.kind) === Pt.IMPLEMENTS) {
    return function getSuggestionsForImplements(e, t, n, r, i) {
      if (t.needsSeparator) {
        return [];
      }
      var a = n.getTypeMap();
      var o = objectValues(a).filter(P);
      var s = o.map(({name: e}) => e);
      var u = new Set;
      runOnlineParser$1(r, (e, t) => {
        var r, a, o, c, l;
        if (t.name) {
          if (t.kind === Pt.INTERFACE_DEF && !s.includes(t.name)) {
            u.add(t.name);
          }
          if (t.kind === Pt.NAMED_TYPE && (null === (r = t.prevState) || void 0 === r ? void 0 : r.kind) === Pt.IMPLEMENTS) {
            if (i.interfaceDef) {
              if (null === (a = i.interfaceDef) || void 0 === a ? void 0 : a.getInterfaces().find(({name: e}) => e === t.name)) {
                return;
              }
              var d = n.getType(t.name);
              var f = null === (o = i.interfaceDef) || void 0 === o ? void 0 : o.toConfig();
              i.interfaceDef = new Z(Object.assign(Object.assign({}, f), {
                interfaces: [ ...f.interfaces, d || new Z({
                  name: t.name,
                  fields: {}
                }) ]
              }));
            } else if (i.objectTypeDef) {
              if (null === (c = i.objectTypeDef) || void 0 === c ? void 0 : c.getInterfaces().find(({name: e}) => e === t.name)) {
                return;
              }
              var v = n.getType(t.name);
              var g = null === (l = i.objectTypeDef) || void 0 === l ? void 0 : l.toConfig();
              i.objectTypeDef = new ee(Object.assign(Object.assign({}, g), {
                interfaces: [ ...g.interfaces, v || new Z({
                  name: t.name,
                  fields: {}
                }) ]
              }));
            }
          }
        }
      });
      var c = i.interfaceDef || i.objectTypeDef;
      var l = ((null == c ? void 0 : c.getInterfaces()) || []).map(({name: e}) => e);
      var d = o.concat([ ...u ].map(e => ({
        name: e
      }))).filter(({name: e}) => e !== (null == c ? void 0 : c.name) && !l.includes(e));
      return hintList$1(e, d.map(e => {
        var t = {
          label: e.name,
          kind: Ct.Interface,
          type: e
        };
        if (null == e ? void 0 : e.description) {
          t.documentation = e.description;
        }
        return t;
      }));
    }(u, c, e, t, g);
  }
  if (d === Pt.SELECTION_SET || d === Pt.FIELD || d === Pt.ALIASED_FIELD) {
    return function getSuggestionsForFieldNames(e, t, n) {
      var r;
      if (t.parentType) {
        var {parentType: i} = t;
        var a = [];
        if ("getFields" in i) {
          a = objectValues(i.getFields());
        }
        if (O(i)) {
          a.push(_);
        }
        if (i === (null === (r = null == n ? void 0 : n.schema) || void 0 === r ? void 0 : r.getQueryType())) {
          a.push(k, L);
        }
        return hintList$1(e, a.map((e, t) => {
          var r;
          var i = {
            sortText: String(t) + e.name,
            label: e.name,
            detail: String(e.type),
            documentation: null !== (r = e.description) && void 0 !== r ? r : void 0,
            deprecated: Boolean(e.deprecationReason),
            isDeprecated: Boolean(e.deprecationReason),
            deprecationReason: e.deprecationReason,
            kind: Ct.Field,
            type: e.type
          };
          if (null == n ? void 0 : n.fillLeafsOnComplete) {
            var a = getInsertText(e);
            if (a) {
              i.insertText = e.name + a;
              i.insertTextFormat = Ge.Snippet;
              i.command = wt;
            }
          }
          return i;
        }));
      }
      return [];
    }(u, g, s);
  }
  if (d === Pt.ARGUMENTS || d === Pt.ARGUMENT && 0 === f) {
    var {argDefs: p} = g;
    if (p) {
      return hintList$1(u, p.map(e => {
        var t;
        return {
          label: e.name,
          insertText: e.name + ": ",
          command: wt,
          detail: String(e.type),
          documentation: null !== (t = e.description) && void 0 !== t ? t : void 0,
          kind: Ct.Variable,
          type: e.type
        };
      }));
    }
  }
  if ((d === Pt.OBJECT_VALUE || d === Pt.OBJECT_FIELD && 0 === f) && g.objectFieldDefs) {
    var m = objectValues(g.objectFieldDefs);
    var E = d === Pt.OBJECT_VALUE ? Ct.Value : Ct.Field;
    return hintList$1(u, m.map(e => {
      var t;
      return {
        label: e.name,
        detail: String(e.type),
        documentation: null !== (t = e.description) && void 0 !== t ? t : void 0,
        kind: E,
        type: e.type
      };
    }));
  }
  if (d === Pt.ENUM_VALUE || d === Pt.LIST_VALUE && 1 === f || d === Pt.OBJECT_FIELD && 2 === f || d === Pt.ARGUMENT && 2 === f) {
    return function getSuggestionsForInputValues(e, t, n, r) {
      var i = j(t.inputType);
      var a = getVariableCompletions(n, r, e).filter(e => e.detail === i.name);
      if (i instanceof Y) {
        return hintList$1(e, i.getValues().map(e => {
          var t;
          return {
            label: e.name,
            detail: String(i),
            documentation: null !== (t = e.description) && void 0 !== t ? t : void 0,
            deprecated: Boolean(e.deprecationReason),
            isDeprecated: Boolean(e.deprecationReason),
            deprecationReason: e.deprecationReason,
            kind: Ct.EnumMember,
            type: i
          };
        }).concat(a));
      }
      if (i === G) {
        return hintList$1(e, a.concat([ {
          label: "true",
          detail: String(G),
          documentation: "Not false.",
          kind: Ct.Variable,
          type: G
        }, {
          label: "false",
          detail: String(G),
          documentation: "Not true.",
          kind: Ct.Variable,
          type: G
        } ]));
      }
      return a;
    }(u, g, t, e);
  }
  if (d === Pt.VARIABLE && 1 === f) {
    var T = j(g.inputType);
    return hintList$1(u, getVariableCompletions(t, e, u).filter(e => e.detail === (null == T ? void 0 : T.name)));
  }
  if (d === Pt.TYPE_CONDITION && 1 === f || d === Pt.NAMED_TYPE && null != v && v.kind === Pt.TYPE_CONDITION) {
    return function getSuggestionsForFragmentTypeConditions(e, t, n, r) {
      var i;
      if (t.parentType) {
        if (B(t.parentType)) {
          var a = Q(t.parentType);
          var o = n.getPossibleTypes(a);
          var s = Object.create(null);
          for (var u of o) {
            for (var c of u.getInterfaces()) {
              s[c.name] = c;
            }
          }
          i = o.concat(objectValues(s));
        } else {
          i = [ t.parentType ];
        }
      } else {
        i = objectValues(n.getTypeMap()).filter(e => O(e) && !e.name.startsWith("__"));
      }
      return hintList$1(e, i.map(e => {
        var t = j(e);
        return {
          label: String(e),
          documentation: (null == t ? void 0 : t.description) || "",
          kind: Ct.Field
        };
      }));
    }(u, g, e);
  }
  if (d === Pt.FRAGMENT_SPREAD && 1 === f) {
    return function getSuggestionsForFragmentSpread$1(e, t, n, r, i) {
      if (!r) {
        return [];
      }
      var a = n.getTypeMap();
      var o = getDefinitionState(e.state);
      var s = function getFragmentDefinitions(e) {
        var t = [];
        runOnlineParser$1(e, (e, n) => {
          if (n.kind === Pt.FRAGMENT_DEFINITION && n.name && n.type) {
            t.push({
              kind: Pt.FRAGMENT_DEFINITION,
              name: {
                kind: x.NAME,
                value: n.name
              },
              selectionSet: {
                kind: Pt.SELECTION_SET,
                selections: []
              },
              typeCondition: {
                kind: Pt.NAMED_TYPE,
                name: {
                  kind: x.NAME,
                  value: n.type
                }
              }
            });
          }
        });
        return t;
      }(r);
      if (i && i.length > 0) {
        s.push(...i);
      }
      var u = s.filter(e => a[e.typeCondition.name.value] && !(o && o.kind === Pt.FRAGMENT_DEFINITION && o.name === e.name.value) && O(t.parentType) && O(a[e.typeCondition.name.value]) && X(n, t.parentType, a[e.typeCondition.name.value]));
      return hintList$1(e, u.map(e => ({
        label: e.name.value,
        detail: String(a[e.typeCondition.name.value]),
        documentation: `fragment ${e.name.value} on ${e.typeCondition.name.value}`,
        kind: Ct.Field,
        type: a[e.typeCondition.name.value]
      })));
    }(u, g, e, t, Array.isArray(i) ? i : (e => {
      var t = [];
      if (e) {
        try {
          J(z(e), {
            FragmentDefinition(e) {
              t.push(e);
            }
          });
        } catch (e) {
          return [];
        }
      }
      return t;
    })(i));
  }
  var h = unwrapType(c);
  if (l === jt.TYPE_SYSTEM && !h.needsAdvance && d === Pt.NAMED_TYPE || d === Pt.LIST_TYPE) {
    if (h.kind === Pt.FIELD_DEF) {
      return hintList$1(u, Object.values(e.getTypeMap()).filter(e => V(e) && !e.name.startsWith("__")).map(e => ({
        label: e.name,
        kind: Ct.Function
      })));
    }
    if (h.kind === Pt.INPUT_VALUE_DEF) {
      return hintList$1(u, Object.values(e.getTypeMap()).filter(e => $(e) && !e.name.startsWith("__")).map(e => ({
        label: e.name,
        kind: Ct.Function
      })));
    }
  }
  if (d === Pt.VARIABLE_DEFINITION && 2 === f || d === Pt.LIST_TYPE && 1 === f || d === Pt.NAMED_TYPE && v && (v.kind === Pt.VARIABLE_DEFINITION || v.kind === Pt.LIST_TYPE || v.kind === Pt.NON_NULL_TYPE)) {
    return function getSuggestionsForVariableDefinition(e, t, n) {
      var r = t.getTypeMap();
      var i = objectValues(r).filter($);
      return hintList$1(e, i.map(e => ({
        label: e.name,
        documentation: e.description,
        kind: Ct.Variable
      })));
    }(u, e);
  }
  if (d === Pt.DIRECTIVE) {
    return function getSuggestionsForDirective(e, t, n, r) {
      var i;
      if (null === (i = t.prevState) || void 0 === i ? void 0 : i.kind) {
        var a = n.getDirectives().filter(e => function canUseDirective(e, t) {
          if (!(null == e ? void 0 : e.kind)) {
            return !1;
          }
          var {kind: n, prevState: r} = e;
          var {locations: i} = t;
          switch (n) {
           case Pt.QUERY:
            return i.includes(K.QUERY);

           case Pt.MUTATION:
            return i.includes(K.MUTATION);

           case Pt.SUBSCRIPTION:
            return i.includes(K.SUBSCRIPTION);

           case Pt.FIELD:
           case Pt.ALIASED_FIELD:
            return i.includes(K.FIELD);

           case Pt.FRAGMENT_DEFINITION:
            return i.includes(K.FRAGMENT_DEFINITION);

           case Pt.FRAGMENT_SPREAD:
            return i.includes(K.FRAGMENT_SPREAD);

           case Pt.INLINE_FRAGMENT:
            return i.includes(K.INLINE_FRAGMENT);

           case Pt.SCHEMA_DEF:
            return i.includes(K.SCHEMA);

           case Pt.SCALAR_DEF:
            return i.includes(K.SCALAR);

           case Pt.OBJECT_TYPE_DEF:
            return i.includes(K.OBJECT);

           case Pt.FIELD_DEF:
            return i.includes(K.FIELD_DEFINITION);

           case Pt.INTERFACE_DEF:
            return i.includes(K.INTERFACE);

           case Pt.UNION_DEF:
            return i.includes(K.UNION);

           case Pt.ENUM_DEF:
            return i.includes(K.ENUM);

           case Pt.ENUM_VALUE:
            return i.includes(K.ENUM_VALUE);

           case Pt.INPUT_DEF:
            return i.includes(K.INPUT_OBJECT);

           case Pt.INPUT_VALUE_DEF:
            switch (null == r ? void 0 : r.kind) {
             case Pt.ARGUMENTS_DEF:
              return i.includes(K.ARGUMENT_DEFINITION);

             case Pt.INPUT_DEF:
              return i.includes(K.INPUT_FIELD_DEFINITION);
            }
          }
          return !1;
        }(t.prevState, e));
        return hintList$1(e, a.map(e => ({
          label: e.name,
          documentation: e.description || "",
          kind: Ct.Function
        })));
      }
      return [];
    }(u, c, e);
  }
  return [];
}

var Ut = " {\n  $1\n}";

var getInsertText = e => {
  var {type: t} = e;
  if (O(t)) {
    return Ut;
  }
  if (W(t) && O(t.ofType)) {
    return Ut;
  }
  if (q(t)) {
    if (O(t.ofType)) {
      return Ut;
    }
    if (W(t.ofType) && O(t.ofType.ofType)) {
      return Ut;
    }
  }
  return null;
};

var getParentDefinition$1 = (e, t) => {
  var n, r, i, a, o, s, u, c, l, d;
  if ((null === (n = e.prevState) || void 0 === n ? void 0 : n.kind) === t) {
    return e.prevState;
  }
  if ((null === (i = null === (r = e.prevState) || void 0 === r ? void 0 : r.prevState) || void 0 === i ? void 0 : i.kind) === t) {
    return e.prevState.prevState;
  }
  if ((null === (s = null === (o = null === (a = e.prevState) || void 0 === a ? void 0 : a.prevState) || void 0 === o ? void 0 : o.prevState) || void 0 === s ? void 0 : s.kind) === t) {
    return e.prevState.prevState.prevState;
  }
  if ((null === (d = null === (l = null === (c = null === (u = e.prevState) || void 0 === u ? void 0 : u.prevState) || void 0 === c ? void 0 : c.prevState) || void 0 === l ? void 0 : l.prevState) || void 0 === d ? void 0 : d.kind) === t) {
    return e.prevState.prevState.prevState.prevState;
  }
};

function getVariableCompletions(e, t, n) {
  var r = null;
  var i;
  var a = Object.create({});
  runOnlineParser$1(e, (e, o) => {
    if ((null == o ? void 0 : o.kind) === Pt.VARIABLE && o.name) {
      r = o.name;
    }
    if ((null == o ? void 0 : o.kind) === Pt.NAMED_TYPE && r) {
      var s = getParentDefinition$1(o, Pt.TYPE);
      if (null == s ? void 0 : s.type) {
        i = t.getType(null == s ? void 0 : s.type);
      }
    }
    if (r && i && !a[r]) {
      a[r] = {
        detail: i.toString(),
        insertText: "$" === n.string ? r : "$" + r,
        label: r,
        type: i,
        kind: Ct.Variable
      };
      r = null;
      i = null;
    }
  });
  return objectValues(a);
}

function getTokenAtPosition(e, t, n = 0) {
  var r = null;
  var i = null;
  var a = null;
  var o = runOnlineParser$1(e, (e, o, s, u) => {
    if (u !== t.line || e.getCurrentPosition() + n < t.character + 1) {
      return;
    }
    r = s;
    i = Object.assign({}, o);
    a = e.current();
    return "BREAK";
  });
  return {
    start: o.start,
    end: o.end,
    string: a || o.string,
    state: i || o.state,
    style: r || o.style
  };
}

function runOnlineParser$1(n, r) {
  var i = n.split("\n");
  var a = t();
  var o = a.startState();
  var s = "";
  var u = new e("");
  for (var c = 0; c < i.length; c++) {
    u = new e(i[c]);
    while (!u.eol()) {
      if ("BREAK" === r(u, o, s = a.token(u, o), c)) {
        break;
      }
    }
    r(u, o, s, c);
    if (!o.kind) {
      o = a.startState();
    }
  }
  return {
    start: u.getStartOfToken(),
    end: u.getCurrentPosition(),
    string: u.current(),
    state: o,
    style: s
  };
}

function getTypeInfo(e, t) {
  var n;
  var r;
  var i;
  var a;
  var o;
  var s;
  var u;
  var c;
  var l;
  var d;
  var f;
  forEachState(t, t => {
    var v;
    switch (t.kind) {
     case Pt.QUERY:
     case "ShortQuery":
      d = e.getQueryType();
      break;

     case Pt.MUTATION:
      d = e.getMutationType();
      break;

     case Pt.SUBSCRIPTION:
      d = e.getSubscriptionType();
      break;

     case Pt.INLINE_FRAGMENT:
     case Pt.FRAGMENT_DEFINITION:
      if (t.type) {
        d = e.getType(t.type);
      }
      break;

     case Pt.FIELD:
     case Pt.ALIASED_FIELD:
      if (!d || !t.name) {
        o = null;
      } else {
        o = l ? getFieldDef(e, l, t.name) : null;
        d = o ? o.type : null;
      }
      break;

     case Pt.SELECTION_SET:
      l = j(d);
      break;

     case Pt.DIRECTIVE:
      i = t.name ? e.getDirective(t.name) : null;
      break;

     case Pt.INTERFACE_DEF:
      if (t.name) {
        u = null;
        f = new Z({
          name: t.name,
          interfaces: [],
          fields: {}
        });
      }
      break;

     case Pt.OBJECT_TYPE_DEF:
      if (t.name) {
        f = null;
        u = new ee({
          name: t.name,
          interfaces: [],
          fields: {}
        });
      }
      break;

     case Pt.ARGUMENTS:
      if (t.prevState) {
        switch (t.prevState.kind) {
         case Pt.FIELD:
          r = o && o.args;
          break;

         case Pt.DIRECTIVE:
          r = i && i.args;
          break;

         case Pt.ALIASED_FIELD:
          var g = null === (v = t.prevState) || void 0 === v ? void 0 : v.name;
          if (!g) {
            r = null;
            break;
          }
          var p = l ? getFieldDef(e, l, g) : null;
          if (!p) {
            r = null;
            break;
          }
          r = p.args;
          break;

         default:
          r = null;
        }
      } else {
        r = null;
      }
      break;

     case Pt.ARGUMENT:
      if (r) {
        for (var m = 0; m < r.length; m++) {
          if (r[m].name === t.name) {
            n = r[m];
            break;
          }
        }
      }
      s = null == n ? void 0 : n.type;
      break;

     case Pt.ENUM_VALUE:
      var E = j(s);
      a = E instanceof Y ? E.getValues().find(e => e.value === t.name) : null;
      break;

     case Pt.LIST_VALUE:
      var T = ne(s);
      s = T instanceof re ? T.ofType : null;
      break;

     case Pt.OBJECT_VALUE:
      var h = j(s);
      c = h instanceof te ? h.getFields() : null;
      break;

     case Pt.OBJECT_FIELD:
      var y = t.name && c ? c[t.name] : null;
      s = null == y ? void 0 : y.type;
      break;

     case Pt.NAMED_TYPE:
      if (t.name) {
        d = e.getType(t.name);
      }
    }
  });
  return {
    argDef: n,
    argDefs: r,
    directiveDef: i,
    enumValue: a,
    fieldDef: o,
    inputType: s,
    objectFieldDefs: c,
    parentType: l,
    type: d,
    interfaceDef: f,
    objectTypeDef: u
  };
}

var jt;

!function(e) {
  e.TYPE_SYSTEM = "TYPE_SYSTEM";
  e.EXECUTABLE = "EXECUTABLE";
}(jt || (jt = {}));

function unwrapType(e) {
  if (e.prevState && e.kind && [ Pt.NAMED_TYPE, Pt.LIST_TYPE, Pt.TYPE, Pt.NON_NULL_TYPE ].includes(e.kind)) {
    return unwrapType(e.prevState);
  }
  return e;
}

function getHoverInformation(e, t, n, r, i) {
  var a = r || getTokenAtPosition(t, n);
  if (!e || !a || !a.state) {
    return "";
  }
  var {kind: o, step: s} = a.state;
  var u = getTypeInfo(e, a.state);
  var c = Object.assign(Object.assign({}, i), {
    schema: e
  });
  if ("Field" === o && 0 === s && u.fieldDef || "AliasedField" === o && 2 === s && u.fieldDef) {
    var l = [];
    renderMdCodeStart(l, c);
    !function renderField(e, t, n) {
      renderQualifiedField(e, t, n);
      renderTypeAnnotation(e, t, n, t.type);
    }(l, u, c);
    renderMdCodeEnd(l, c);
    renderDescription(l, c, u.fieldDef);
    return l.join("").trim();
  }
  if ("Directive" === o && 1 === s && u.directiveDef) {
    var d = [];
    renderMdCodeStart(d, c);
    renderDirective(d, u);
    renderMdCodeEnd(d, c);
    renderDescription(d, c, u.directiveDef);
    return d.join("").trim();
  }
  if ("Argument" === o && 0 === s && u.argDef) {
    var f = [];
    renderMdCodeStart(f, c);
    !function renderArg(e, t, n) {
      if (t.directiveDef) {
        renderDirective(e, t);
      } else if (t.fieldDef) {
        renderQualifiedField(e, t, n);
      }
      if (!t.argDef) {
        return;
      }
      var {name: r} = t.argDef;
      text(e, "(");
      text(e, r);
      renderTypeAnnotation(e, t, n, t.inputType);
      text(e, ")");
    }(f, u, c);
    renderMdCodeEnd(f, c);
    renderDescription(f, c, u.argDef);
    return f.join("").trim();
  }
  if ("EnumValue" === o && u.enumValue && "description" in u.enumValue) {
    var v = [];
    renderMdCodeStart(v, c);
    !function renderEnumValue(e, t, n) {
      if (!t.enumValue) {
        return;
      }
      var {name: r} = t.enumValue;
      renderType(e, t, n, t.inputType);
      text(e, ".");
      text(e, r);
    }(v, u, c);
    renderMdCodeEnd(v, c);
    renderDescription(v, c, u.enumValue);
    return v.join("").trim();
  }
  if ("NamedType" === o && u.type && "description" in u.type) {
    var g = [];
    renderMdCodeStart(g, c);
    renderType(g, u, c, u.type);
    renderMdCodeEnd(g, c);
    renderDescription(g, c, u.type);
    return g.join("").trim();
  }
  return "";
}

function renderMdCodeStart(e, t) {
  if (t.useMarkdown) {
    text(e, "```graphql\n");
  }
}

function renderMdCodeEnd(e, t) {
  if (t.useMarkdown) {
    text(e, "\n```");
  }
}

function renderQualifiedField(e, t, n) {
  if (!t.fieldDef) {
    return;
  }
  var r = t.fieldDef.name;
  if ("__" !== r.slice(0, 2)) {
    renderType(e, t, n, t.parentType);
    text(e, ".");
  }
  text(e, r);
}

function renderDirective(e, t, n) {
  if (!t.directiveDef) {
    return;
  }
  text(e, "@" + t.directiveDef.name);
}

function renderTypeAnnotation(e, t, n, r) {
  text(e, ": ");
  renderType(e, t, n, r);
}

function renderType(e, t, n, r) {
  if (!r) {
    return;
  }
  if (r instanceof ie) {
    renderType(e, t, n, r.ofType);
    text(e, "!");
  } else if (r instanceof re) {
    text(e, "[");
    renderType(e, t, n, r.ofType);
    text(e, "]");
  } else {
    text(e, r.name);
  }
}

function renderDescription(e, t, n) {
  if (!n) {
    return;
  }
  var r = "string" == typeof n.description ? n.description : null;
  if (r) {
    text(e, "\n\n");
    text(e, r);
  }
  !function renderDeprecation(e, t, n) {
    if (!n) {
      return;
    }
    var r = n.deprecationReason || null;
    if (!r) {
      return;
    }
    text(e, "\n\n");
    text(e, "Deprecated: ");
    text(e, r);
  }(e, 0, n);
}

function text(e, t) {
  e.push(t);
}

class Cursor {
  constructor(e, t) {
    this.line = e;
    this.character = t;
  }
  setLine(e) {
    this.line = e;
  }
  setCharacter(e) {
    this.character = e;
  }
  lessThanOrEqualTo(e) {
    return this.line < e.line || this.line === e.line && this.character <= e.character;
  }
}

var getToken = (r, i) => {
  if (!n.isTemplateLiteral(r) && !n.isStringLiteralLike(r)) {
    return;
  }
  var a = r.getText().slice(1, -1).split("\n");
  var o = t();
  var s = o.startState();
  var u = r.getStart() + 1;
  var c = void 0;
  var l = void 0;
  for (var d = 0; d < a.length; d++) {
    if (c) {
      continue;
    }
    var f = u - 1;
    var v = new e(a[d] + "\n");
    while (!v.eol()) {
      var g = o.token(v, s);
      var p = v.current();
      if (f + v.getStartOfToken() + 1 <= i && f + v.getCurrentPosition() >= i) {
        c = l ? l : {
          line: d,
          start: v.getStartOfToken() + 1,
          end: v.getCurrentPosition(),
          string: p,
          state: s,
          tokenKind: g
        };
        break;
      } else if ("on" === p) {
        l = {
          line: d,
          start: v.getStartOfToken() + 1,
          end: v.getCurrentPosition(),
          string: p,
          state: s,
          tokenKind: g
        };
      } else if ("." === p || ".." === p) {
        l = {
          line: d,
          start: v.getStartOfToken() + 1,
          end: v.getCurrentPosition(),
          string: p,
          state: s,
          tokenKind: g
        };
      } else {
        l = void 0;
      }
    }
    u += a[d].length + 1;
  }
  return c;
};

function hintList(e, t) {
  return function filterAndSortList(e, t) {
    if (!t) {
      return filterNonEmpty(e, e => !e.isDeprecated);
    }
    var n = e.map(e => ({
      proximity: getProximity(normalizeText(e.label), t),
      entry: e
    }));
    return filterNonEmpty(filterNonEmpty(n, e => e.proximity <= 2), e => !e.entry.isDeprecated).sort((e, t) => (e.entry.isDeprecated ? 1 : 0) - (t.entry.isDeprecated ? 1 : 0) || e.proximity - t.proximity || e.entry.label.length - t.entry.label.length).map(e => e.entry);
  }(t, normalizeText(e.string));
}

function filterNonEmpty(e, t) {
  var n = e.filter(t);
  return 0 === n.length ? e : n;
}

function normalizeText(e) {
  return e.toLowerCase().replace(/\W/g, "");
}

function getProximity(e, t) {
  var n = function lexicalDistance(e, t) {
    var n;
    var r;
    var i = [];
    var a = e.length;
    var o = t.length;
    for (n = 0; n <= a; n++) {
      i[n] = [ n ];
    }
    for (r = 1; r <= o; r++) {
      i[0][r] = r;
    }
    for (n = 1; n <= a; n++) {
      for (r = 1; r <= o; r++) {
        var s = e[n - 1] === t[r - 1] ? 0 : 1;
        i[n][r] = Math.min(i[n - 1][r] + 1, i[n][r - 1] + 1, i[n - 1][r - 1] + s);
        if (n > 1 && r > 1 && e[n - 1] === t[r - 2] && e[n - 2] === t[r - 1]) {
          i[n][r] = Math.min(i[n][r], i[n - 2][r - 2] + s);
        }
      }
    }
    return i[a][o];
  }(t, e);
  if (e.length > t.length) {
    n -= e.length - t.length - 1;
    n += 0 === e.indexOf(t) ? 0 : .5;
  }
  return n;
}

function getGraphQLCompletions(e, t, v, g) {
  var p = g.config.templateIsCallExpression ?? !0;
  var m = g.languageService.getProgram()?.getTypeChecker();
  var E = r(g, e);
  if (!E) {
    return;
  }
  var T = i(E, t);
  if (!T) {
    return;
  }
  T = p ? a(T) : o(T);
  var h, y, S;
  if (p && s(T, m)) {
    var I = u(T, m);
    S = getSchemaForName(v, I);
    var b = getToken(T.arguments[0], t);
    if (!S || !b || "." === b.string || ".." === b.string) {
      return;
    }
    h = `${T.arguments[0].getText().slice(1, -1)}\n${c(T, g).map(e => l(e)).join("\n")}`;
    y = new Cursor(b.line, b.start - 1);
  } else if (!p && d(T)) {
    var N = getToken(T.template, t);
    if (!N || !v.current || "." === N.string || ".." === N.string) {
      return;
    }
    var {combinedText: D, resolvedSpans: A} = f(T, e, g);
    var F = A.filter(e => e.original.start < t && e.original.start + e.original.length < t).reduce((e, t) => e + (t.lines - 1), 0);
    N.line = N.line + F;
    h = D;
    y = new Cursor(N.line, N.start - 1);
    S = v.current.schema;
  } else {
    return;
  }
  var [k, L] = function getSuggestionsInternal(e, t, n) {
    var r = getTokenAtPosition(t, n);
    var i = [];
    try {
      i = z(t, {
        noLocation: !0
      }).definitions.filter(e => e.kind === x.FRAGMENT_DEFINITION);
    } catch (e) {}
    var a = "on" === r.string && "TypeCondition" === r.state.kind;
    var o = getAutocompleteSuggestions(e, t, n, a ? {
      ...r,
      state: {
        ...r.state,
        step: 1
      },
      type: null
    } : void 0);
    var s = !a ? function getSuggestionsForFragmentSpread(e, t, n, r, i) {
      if (!r) {
        return [];
      }
      var a = n.getTypeMap();
      var o = getDefinitionState(e.state);
      return hintList(e, i.filter(e => a[e.typeCondition.name.value] && !(o && o.kind === Pt.FRAGMENT_DEFINITION && o.name === e.name.value) && O(t.parentType) && O(a[e.typeCondition.name.value]) && X(n, t.parentType, a[e.typeCondition.name.value])).map(e => ({
        label: e.name.value,
        detail: String(a[e.typeCondition.name.value]),
        documentation: `fragment ${e.name.value} on ${e.typeCondition.name.value}`,
        kind: Ct.Field,
        type: a[e.typeCondition.name.value]
      })));
    }(r, getTypeInfo(e, r.state), e, t, i) : [];
    var u = "Invalid" === r.state.kind ? r.state.prevState : r.state;
    var c = getParentDefinition(r.state, Pt.FIELD)?.name;
    if (u && c) {
      var {kind: l} = u;
      if (l === Pt.ARGUMENTS || l === Pt.ARGUMENT) {
        var d = new Set;
        runOnlineParser(t, (e, t) => {
          if (t.kind === Pt.ARGUMENT) {
            var n = getParentDefinition(t, Pt.FIELD);
            if (c && t.name && n?.name === c) {
              d.add(t.name);
            }
          }
        });
        o = o.filter(e => !d.has(e.label));
      }
      if (l === Pt.SELECTION_SET || l === Pt.FIELD || l === Pt.ALIASED_FIELD) {
        var f = new Set;
        var v = getUsedFragments(t, c);
        runOnlineParser(t, (e, t) => {
          if (t.kind === Pt.FIELD || t.kind === Pt.ALIASED_FIELD) {
            var n = getParentDefinition(t, Pt.FIELD);
            if (n && n.name === c && t.name) {
              f.add(t.name);
            }
          }
        });
        o = o.filter(e => !f.has(e.label));
        s = s.filter(e => !v.has(e.label));
      }
      if (l === Pt.FRAGMENT_SPREAD) {
        var g = getUsedFragments(t, c);
        o = o.filter(e => !g.has(e.label));
        s = s.filter(e => !g.has(e.label));
      }
    }
    return [ o, s ];
  }(S, h, y);
  return {
    isGlobalCompletion: !1,
    isMemberCompletion: !1,
    isNewIdentifierLocation: !1,
    entries: [ ...k.map(e => ({
      ...e,
      kind: n.ScriptElementKind.variableElement,
      name: e.label,
      kindModifiers: "declare",
      sortText: e.sortText || "0",
      labelDetails: {
        detail: e.type ? " " + e.type?.toString() : void 0,
        description: e.documentation
      }
    })), ...L.map(e => ({
      ...e,
      kind: n.ScriptElementKind.variableElement,
      name: e.label,
      insertText: "..." + e.label,
      kindModifiers: "declare",
      sortText: "0",
      labelDetails: {
        description: e.documentation
      }
    })) ]
  };
}

function getUsedFragments(e, t) {
  var n = new Set;
  runOnlineParser(e, (e, r) => {
    if (r.kind === Pt.FRAGMENT_SPREAD && r.name) {
      var i = getParentDefinition(r, Pt.FIELD);
      if (t && i?.name === t) {
        n.add(r.name);
      }
    }
  });
  return n;
}

function getParentDefinition(e, t) {
  if (e.prevState?.kind === t) {
    return e.prevState;
  }
  if (e.prevState?.prevState?.kind === t) {
    return e.prevState.prevState;
  }
  if (e.prevState?.prevState?.prevState?.kind === t) {
    return e.prevState.prevState.prevState;
  }
  if (e.prevState?.prevState?.prevState?.prevState?.kind === t) {
    return e.prevState.prevState.prevState.prevState;
  }
}

function runOnlineParser(n, r) {
  var i = n.split("\n");
  var a = t();
  var o = a.startState();
  var s = "";
  var u = new e("");
  for (var c = 0; c < i.length; c++) {
    u = new e(i[c]);
    while (!u.eol()) {
      if ("BREAK" === r(u, o, s = a.token(u, o), c)) {
        break;
      }
    }
    r(u, o, s, c);
    if (!o.kind) {
      o = a.startState();
    }
  }
  return {
    start: u.getStartOfToken(),
    end: u.getCurrentPosition(),
    string: u.current(),
    state: o,
    style: s
  };
}

var isPositionInRange = (e, t, n) => e >= t && e < n;

var makeDefinitionInfo = e => ({
  fileName: e.fileName,
  textSpan: {
    start: e.start,
    length: e.length
  },
  kind: n.ScriptElementKind.memberVariableElement,
  name: e.fieldName,
  containerKind: n.ScriptElementKind.interfaceElement,
  containerName: e.parentName
});

var getNameText = e => {
  if (n.isIdentifier(e) || n.isStringLiteralLike(e)) {
    return e.text;
  }
  if (n.isNumericLiteral(e)) {
    return e.text;
  }
  return;
};

var getTadaCacheMatch = e => {
  var t = [];
  var r = e;
  var i;
  while (r) {
    if (n.isPropertySignature(r)) {
      if (r.parent && n.isInterfaceDeclaration(r.parent) && "setupCache" === r.parent.name.text && n.isStringLiteralLike(r.name)) {
        i = r;
        break;
      }
      var a = getNameText(r.name);
      if (!a) {
        return;
      }
      t.push(a);
    }
    r = r.parent;
  }
  if (!i || !n.isStringLiteralLike(i.name)) {
    return;
  }
  var o = i.type;
  if (!o || !n.isTypeReferenceNode(o)) {
    return;
  }
  if ("TadaDocumentNode" !== (e => {
    var t = e.typeName;
    return n.isIdentifier(t) ? t.text : t.right.text;
  })(o)) {
    return;
  }
  var s = o.typeArguments?.[0];
  if (!s || !(u = s, c = e, c.getStart() >= u.getStart() && c.getEnd() <= u.getEnd())) {
    return;
  }
  var u, c;
  t.reverse();
  if (!t.length) {
    return;
  }
  return {
    documentText: i.name.text,
    responsePath: t
  };
};

var findTadaCacheMatch = (e, t) => {
  var n = t.getSourceFile(e.fileName);
  if (!n) {
    return;
  }
  var r = i(n, e.textSpan.start);
  return r && getTadaCacheMatch(r);
};

var findDocumentCallInSource = (e, t, n, r) => {
  if (e.isDeclarationFile) {
    return;
  }
  if (!e.getText().includes(t)) {
    return;
  }
  var {nodes: i} = v(e, r, {
    searchExternal: !1,
    collectFragments: !1
  });
  for (var a of i) {
    if (a.node.text !== t) {
      continue;
    }
    if (n && a.schema !== n) {
      continue;
    }
    return {
      fileName: e.fileName,
      node: a.node
    };
  }
};

var findDocumentCall = (e, t, n, r, i) => {
  var a = r.getSourceFile(n);
  if (a) {
    var o = findDocumentCallInSource(a, e, t, i);
    if (o) {
      return o;
    }
  }
  for (var s of r.getSourceFiles()) {
    if (s === a) {
      continue;
    }
    var u = findDocumentCallInSource(s, e, t, i);
    if (u) {
      return u;
    }
  }
};

var getSchemaNameForTurboFile = (e, t) => {
  var normalize = e => y.normalize(e).toLowerCase();
  var n = normalize(e);
  for (var [r, i] of t.turboLocations) {
    if (normalize(i) === n) {
      return r;
    }
  }
  return null;
};

var findFieldByResponsePath = (e, t) => {
  var n;
  try {
    n = z(e);
  } catch (e) {
    return;
  }
  var r = new Map;
  for (var i of n.definitions) {
    if (i.kind === x.FRAGMENT_DEFINITION) {
      r.set(i.name.value, i);
    }
  }
  var findInSelectionSet = (e, t, n = new Set) => {
    var [i, ...a] = t;
    if (!i) {
      return;
    }
    for (var o of e.selections) {
      if (o.kind === x.FIELD) {
        if ((o.alias ? o.alias.value : o.name.value) !== i) {
          continue;
        }
        if (!a.length) {
          return o;
        }
        if (!o.selectionSet) {
          return;
        }
        var s = findInSelectionSet(o.selectionSet, a);
        if (s) {
          return s;
        }
      } else if (o.kind === x.INLINE_FRAGMENT) {
        var u = findInSelectionSet(o.selectionSet, t, n);
        if (u) {
          return u;
        }
      } else if (o.kind === x.FRAGMENT_SPREAD) {
        var c = o.name.value;
        if (n.has(c)) {
          continue;
        }
        var l = r.get(c);
        if (!l) {
          continue;
        }
        n.add(c);
        var d = findInSelectionSet(l.selectionSet, t, n);
        if (d) {
          return d;
        }
      }
    }
  };
  for (var a of n.definitions) {
    if (a.kind === x.OPERATION_DEFINITION || a.kind === x.FRAGMENT_DEFINITION) {
      var o = findInSelectionSet(a.selectionSet, t);
      if (o) {
        return o;
      }
    }
  }
};

var getDefinitionForDocumentField = (e, t, c, l) => {
  var f = l.config.templateIsCallExpression ?? !0;
  var v = l.languageService.getProgram()?.getTypeChecker();
  var g = r(l, e);
  if (!g) {
    return;
  }
  var p = i(g, t);
  if (!p) {
    return;
  }
  p = f ? a(p) : o(p);
  var m;
  var E = null;
  var T;
  if (f && s(p, v)) {
    E = u(p, v);
    T = getSchemaForName(c, E);
    m = p.arguments[0];
  } else if (!f && d(p)) {
    if (!n.isNoSubstitutionTemplateLiteral(p.template)) {
      return;
    }
    T = c.current?.schema;
    m = p.template;
  } else {
    return;
  }
  if (!T || !n.isStringLiteralLike(m)) {
    return;
  }
  var h = m.getStart() + 1;
  var y = t - h;
  if (y < 0 || y >= m.text.length) {
    return;
  }
  var S = ((e, t, n) => {
    var r;
    try {
      r = z(t);
    } catch (e) {
      return;
    }
    var i;
    var a = new ae(e);
    J(r, oe(a, {
      Field: {
        enter(e) {
          var t = e.name.loc;
          var r = e.alias?.loc;
          var o = t && isPositionInRange(n, t.start, t.end) ? t : r && isPositionInRange(n, r.start, r.end) ? r : null;
          if (!o) {
            return;
          }
          var s = a.getParentType();
          if (!s) {
            return;
          }
          i = {
            field: e,
            parentType: s,
            boundStart: o.start,
            boundLength: o.end - o.start
          };
          return !1;
        }
      }
    }));
    return i;
  })(T, m.text, y);
  if (!S) {
    return;
  }
  var I = ((e, t, n, r) => {
    var i = e.sourceLocations.get(t);
    if (!i) {
      return;
    }
    if (!C(n) && !P(n)) {
      return;
    }
    var a = n.getFields()[r];
    var o = a?.astNode?.name.loc;
    if (!a || !o) {
      return;
    }
    return {
      fileName: i,
      fieldName: r,
      parentName: n.name,
      start: o.start,
      length: o.end - o.start
    };
  })(c, E, S.parentType, S.field.name.value);
  if (!I) {
    return;
  }
  return {
    textSpan: {
      start: h + S.boundStart,
      length: S.boundLength
    },
    definitions: [ makeDefinitionInfo(I) ]
  };
};

var getDefinitionForTadaResultField = (e, t, r, i) => {
  if (!i?.length) {
    return;
  }
  var a = r.languageService.getProgram();
  if (!a) {
    return;
  }
  var o = !1;
  for (var s of i) {
    var u = findTadaCacheMatch(s, a);
    if (!u) {
      continue;
    }
    o = !0;
    var c = getSchemaNameForTurboFile(s.fileName, t);
    var l = findDocumentCall(u.documentText, c, e, a, r);
    if (!l) {
      continue;
    }
    var d = findFieldByResponsePath(u.documentText, u.responsePath);
    var f = (d?.alias || d?.name)?.loc;
    if (!f) {
      continue;
    }
    return [ {
      fileName: l.fileName,
      textSpan: {
        start: l.node.getStart() + 1 + f.start,
        length: f.end - f.start
      },
      kind: n.ScriptElementKind.memberVariableElement,
      name: u.responsePath[u.responsePath.length - 1],
      containerKind: n.ScriptElementKind.scriptElement,
      containerName: ""
    } ];
  }
  return o ? [] : void 0;
};

function create(e) {
  var logger = t => e.project.projectService.logger.info(`[GraphQLSP] ${t}`);
  var t = e.config;
  logger("config: " + JSON.stringify(t));
  logger("Setting up the GraphQL Plugin");
  if (t.template) {
    p.add(t.template);
  }
  var c = function createBasicDecorator(e) {
    var t = Object.create(null);
    var _loop = function() {
      var r = e.languageService[n];
      t[n] = (...t) => r.apply(e.languageService, t);
    };
    for (var n of Object.keys(e.languageService)) {
      _loop();
    }
    return t;
  }(e);
  var l = loadSchema(e, logger);
  var guard = (e, t, n) => {
    try {
      return n();
    } catch (n) {
      logger(`Unexpected error in ${e}: ${n}`);
      return t;
    }
  };
  c.getSemanticDiagnostics = t => {
    var n = e.languageService.getSemanticDiagnostics(t);
    return guard("getSemanticDiagnostics", n, () => {
      l.checkStale();
      if (n.some(e => m.includes(e.code))) {
        return n;
      }
      var r = E(t, l, e);
      return r ? [ ...r, ...n ] : n;
    });
  };
  c.getCompletionsAtPosition = (t, n, r) => {
    var i = guard("getCompletionsAtPosition", void 0, () => getGraphQLCompletions(t, n, l, e));
    if (i && i.entries.length) {
      return i;
    } else {
      return e.languageService.getCompletionsAtPosition(t, n, r) || {
        isGlobalCompletion: !1,
        isMemberCompletion: !1,
        isNewIdentifierLocation: !1,
        entries: []
      };
    }
  };
  c.getEditsForRefactor = (t, n, r, i, a, o, s) => {
    var u = e.languageService.getEditsForRefactor(t, n, r, i, a, o, s);
    var c = guard("getEditsForRefactor", void 0, () => T(t, "number" == typeof r ? r : r.pos, e));
    if (!c) {
      return u;
    }
    return {
      edits: [ {
        fileName: t,
        textChanges: [ {
          newText: c.replacement,
          span: c.span
        } ]
      } ]
    };
  };
  c.getApplicableRefactors = (t, n, r, i, a, o) => {
    var s = e.languageService.getApplicableRefactors(t, n, r, i, a, o);
    if (guard("getApplicableRefactors", void 0, () => T(t, "number" == typeof n ? n : n.pos, e))) {
      return [ {
        name: "GraphQL",
        description: "Operations specific to gql.tada!",
        actions: [ {
          name: "Insert document-id",
          description: "Generate a document-id for your persisted-operation, by default a SHA256 hash."
        } ],
        inlineable: !0
      }, ...s ];
    } else {
      return s;
    }
  };
  c.getDefinitionAtPosition = (t, n) => {
    var r = e.languageService.getDefinitionAtPosition(t, n);
    var i = guard("getDefinitionAtPosition", void 0, () => function getGraphQLDefinitionAtPosition(e, t, n, r, i) {
      var a = getDefinitionForDocumentField(e, t, n, r);
      if (a?.definitions?.length) {
        return a.definitions;
      }
      return getDefinitionForTadaResultField(e, n, r, i);
    }(t, n, l, e, r));
    return i || r;
  };
  c.getDefinitionAndBoundSpan = (t, n) => {
    var r = e.languageService.getDefinitionAndBoundSpan(t, n);
    var i = guard("getDefinitionAndBoundSpan", void 0, () => function getGraphQLDefinitionAndBoundSpan(e, t, n, r, i) {
      var a = getDefinitionForDocumentField(e, t, n, r);
      if (a?.definitions?.length) {
        return a;
      }
      var o = getDefinitionForTadaResultField(e, n, r, i?.definitions);
      if (!o) {
        return;
      }
      return {
        textSpan: i?.textSpan || {
          start: t,
          length: 1
        },
        definitions: o
      };
    }(t, n, l, e, r));
    return i || r;
  };
  c.getQuickInfoAtPosition = (...t) => {
    var [c, v] = t;
    var g = guard("getQuickInfoAtPosition", void 0, () => function getGraphQLQuickInfo(e, t, c, l) {
      var v = l.config.templateIsCallExpression ?? !0;
      var g = l.languageService.getProgram()?.getTypeChecker();
      var p = r(l, e);
      if (!p) {
        return;
      }
      var m = i(p, t);
      if (!m) {
        return;
      }
      m = v ? a(m) : o(m);
      var E, T, h;
      if (v && s(m, g)) {
        var y = l.languageService.getProgram()?.getTypeChecker();
        var S = u(m, y);
        h = getSchemaForName(c, S);
        var I = getToken(m.arguments[0], t);
        if (!h || !I) {
          return;
        }
        T = m.arguments[0].getText();
        E = new Cursor(I.line, I.start - 1);
      } else if (!v && d(m)) {
        var b = getToken(m.template, t);
        if (!b || !c.current) {
          return;
        }
        var {combinedText: N, resolvedSpans: D} = f(m, e, l);
        var A = D.filter(e => e.original.start < t && e.original.start + e.original.length < t).reduce((e, t) => e + (t.lines - 1), 0);
        b.line = b.line + A;
        T = N;
        E = new Cursor(b.line, b.start - 1);
        h = c.current.schema;
      } else {
        return;
      }
      var F = getHoverInformation(h, T, E);
      return {
        kind: n.ScriptElementKind.label,
        textSpan: {
          start: t,
          length: 1
        },
        kindModifiers: "text",
        documentation: Array.isArray(F) ? F.map(e => ({
          kind: "text",
          text: e
        })) : [ {
          kind: "text",
          text: F
        } ]
      };
    }(c, v, l, e));
    if (g) {
      return g;
    }
    return e.languageService.getQuickInfoAtPosition(...t);
  };
  logger("proxy: " + JSON.stringify(c));
  return c;
}

var init = e => {
  g(e);
  return {
    create
  };
};

export { init as default };
